pkg/autohealing/cloudprovider/openstack/provider.go

```diff
@@ -250,6 +250,61 @@ func (provider OpenStackCloudProvider) waitForServerDetachVolumes(serverID strin
 	return rootVolumeID, err
 }
 
+// FirstTimeRepair Handle the first time repair for a node
+// 1) If the node is the first time in error, reboot it
+// 2) If the node is not the first time in error, check if the last reboot time is in provider.Config.RebuildDelayAfterReboot
+func (provider OpenStackCloudProvider) FirstTimeRepair(n healthcheck.NodeInfo, serverID string, firstTimeRebootNodes map[string]healthcheck.NodeInfo) (bool, error) {
+	var firsttimeUnhealthy = true
+	for id := range unHealthyNodes {
+		log.V(5).Infof("comparing server ID %s with known broken ID %s", serverID, id)
+		if id == serverID {
+			log.Infof("it is NOT the first time that node %s is being repaired", serverID)
+			firsttimeUnhealthy = false
+			break
+		}
+	}
+
+	var processed = false
+	if firsttimeUnhealthy == true {
+		unHealthyNodes[serverID] = n
+		log.Infof("rebooting node %s to repair it", serverID)
+
+		if res := servers.Reboot(provider.Nova, serverID, servers.RebootOpts{Type: servers.SoftReboot}); res.Err != nil {
+			// Usually it means the node is being rebooted
+			log.Warningf("failed to reboot node %s, error: %v", serverID, res.Err)
+		} else {
+			// Uncordon the node
+			if n.IsWorker == true {
+				nodeName := n.KubeNode.Name
+				newNode := n.KubeNode.DeepCopy()
+				newNode.Spec.Unschedulable = false
+				if _, err := provider.KubeClient.CoreV1().Nodes().Update(context.TODO(), newNode, metav1.UpdateOptions{}); err != nil {
+					log.Errorf("Failed to cordon node %s, error: %v", nodeName, err)
+				} else {
+					log.Infof("Node %s is cordoned", nodeName)
+				}
+			}
+
+			n.RebootAt = time.Now()
+			firstTimeRebootNodes[serverID] = n
+			unHealthyNodes[serverID] = n
+			log.Infof("Node %s has been rebooted at %s", serverID, n.RebootAt)
+		}
+		processed = true
+	} else {
+		n.RebootAt = unHealthyNodes[serverID].RebootAt
+		// If the node was rebooted within mins (defined by RepairDelayAfterReboot), ignore it
+		log.Infof("Node %s is found in unhealthy again which was rebooted at %s", serverID, n.RebootAt)
+		if n.RebootAt.After(time.Now().Add(-1 * provider.Config.RebuildDelayAfterReboot)) {
+			log.Infof("Node %s is found in unhealthy again, but we're going to defer the repair because it maybe in reboot process", serverID)
+			firstTimeRebootNodes[serverID] = n
+			processed = true
+		}
```

Comment:
If it's rebooted for a while ago but still presented as un healthy, we just delete it

Reply:
@ricolin  What do you mean "we just delete it"? We cannot just delete the node, since it's being rebooted.

Reply:
Just noted for the behavior, so correct me if I misinterpret it

I mean we will not found node in firstTimeRebootNodes [1] if it's not firsttimeUnhealthy and already passed delay mins since last reboot. And that node will be deleted at line 471

[1] https://github.com/kubernetes/cloud-provider-openstack/pull/1530/commits/3d06649e21aa6a19c59477640b970efe6ec96736#diff-9846ce0766626342e9df737eff34e6aba6b203aa80e0ba8fa021ac26ddbad785R467

Reply:
Yes, if the node is not firsttimeUnhealthy and already passed delayed mins, then it will be deleted from the k8s PoV.